File: pptx-service/renderers/chart_bar.py
from builder import add_bg, add_textbox, add_rect, add_line
from theme import ThemeColors


def render(slide, data: dict, t: ThemeColors) -> None:
    add_bg(slide, t.bg)
    heading = data.get("heading", "")
    unit = data.get("unit", "")
    bars = data.get("bars", [])
    highlight = data.get("highlight", "")
    source = data.get("source", "")

    add_textbox(slide, 0.83, 0.4, 11.0, 1.0, heading,
                t.font_display, 44, t.text, bold=True)
    if unit:
        add_textbox(slide, 0.83, 1.3, 4.0, 0.4, f"Unit: {unit}",
                    t.font_body, 16, t.muted)

    bars = bars[:8]
    if not bars:
        return
    max_val = max((b.get("value", 0) for b in bars), default=1)
    if max_val == 0:
        max_val = 1

    chart_x = 2.0
    chart_w = 10.5
    chart_top = 1.8
    chart_h = 5.0
    bar_h = chart_h / len(bars) * 0.65
    gap_h = chart_h / len(bars) * 0.35

    for i, bar in enumerate(bars):
        y = chart_top + i * (bar_h + gap_h)
        val = bar.get("value", 0)
        label = bar.get("label", "")
        is_highlight = label == highlight
        fill = t.accent if is_highlight else t.muted
        pct = val / max_val
        bar_w = max(chart_w * pct, 0.05)

        add_textbox(slide, 0.2, y, 1.7, bar_h, label,
                    t.font_body, 18,
                    t.text if is_highlight else t.muted,
                    bold=is_highlight, align="right")
        add_rect(slide, chart_x, y, bar_w, bar_h, fill_hex=fill)
        add_textbox(slide, chart_x + bar_w + 0.08, y, 1.5, bar_h,
                    f"{val:g}", t.font_body, 18, t.text)

    if source:
        add_textbox(slide, 0.83, 7.1, 10.0, 0.35, f"Source: {source}",
                    t.font_body, 14, t.muted)

File: lib/pptx/slides/chart-bar.ts
import type { ChartBarSlide } from '@/lib/types';
import {
  px, py, SLIDE_W,
  setBackground, addRect, addTextBox, addHeading, addPageNumber,
  type PptxTheme,
} from '../helpers';
import type PptxGenJS from 'pptxgenjs';
type PSlide = ReturnType<InstanceType<typeof PptxGenJS>['addSlide']>;

export function render(slide: PSlide, data: ChartBarSlide, t: PptxTheme, n: number, total: number): void {
  setBackground(slide, '#' + t.bg);
  addHeading(slide, data.heading, t, py(60));

  const bars = data.bars.slice(0, 8);
  const maxVal = Math.max(...bars.map(b => b.value), 1);
  const barAreaLeft = t.paddingIn + px(180);
  const barAreaW = SLIDE_W - barAreaLeft - t.paddingIn - px(120);
  const rowH = py(720) / bars.length;
  const barH = rowH * 0.55;
  const startY = py(200);

  bars.forEach((bar, i) => {
    const y = startY + i * rowH;
    const isHL = bar.label === data.highlight;
    const barW = (bar.value / maxVal) * barAreaW;

    addTextBox(slide, bar.label, {
      x: t.paddingIn, y: y + rowH * 0.1, w: px(170),
      fontSize: t.captionSize, color: '#' + t.muted, fontFace: t.fontBody, align: 'right',
    });
    addRect(slide, barAreaLeft, y + (rowH - barH) / 2, barW, barH, '#' + (isHL ? t.accent : t.rule));
    addTextBox(slide, String(bar.value), {
      x: barAreaLeft + barW + px(8), y: y + (rowH - barH) / 2, w: px(100),
      fontSize: t.bodySize, color: '#' + t.text, fontFace: t.fontBody, bold: isHL,
    });
  });

  if (data.source) {
    addTextBox(slide, data.source, {
      x: t.paddingIn, y: py(940), w: SLIDE_W - 2 * t.paddingIn,
      fontSize: t.captionSize, color: '#' + t.muted, fontFace: t.fontBody,
    });
  }

  addPageNumber(slide, n, total, t);
}
